src/allmydata/test/web/test_web.py

```diff
@@ -657,55 +645,65 @@ def should302(self, res, which):
                       (which, res))
 
 
-
-class MultiFormatPageTests(TrialTestCase):
+class MultiFormatResourceTests(TrialTestCase):
     """
-    Tests for ``MultiFormatPage``.
+    Tests for ``MultiFormatResource``.
     """
+
     def resource(self):
         """
-        Create and return an instance of a ``MultiFormatPage`` subclass with two
-        formats: ``a`` and ``b``.
+        Create and return an instance of a ``MultiFormatResource`` subclass
+        with a default HTML format, and two custom formats: ``a`` and ``b``.
         """
-        class Content(MultiFormatPage):
-            docFactory = stan("doc factory")
+        class Content(MultiFormatResource):
+
+            def render_HTML(self, req):
+                return "html"
 
             def render_A(self, req):
                 return "a"
 
             def render_B(self, req):
                 return "b"
+
         return Content()
 
 
     def render(self, resource, **query_args):
         """
-        Render a Nevow ``Page`` against a request with the given query arguments.
+        Render a ``Resource`` against a request with the given query arguments.
 
         :param resource: The Nevow resource to render.
 
         :param query_args: The query arguments to put into the request being
             rendered.  A mapping from ``bytes`` to ``list`` of ``bytes``.
 
-        :return: The rendered response body as ``bytes``.
+        :return: The rendered response body as ``str``.
```

Comment:
Why this change?  What does `str` mean here?  On Python 2 it's exactly the same as `bytes` so this seems like a non-change that will just confuse the Python 3 porting effort.

Reply:
I did a `type()` on the return value and it printed `str`.  Will drop it.

Reply:
Restored the old docstring in d5b5177.

Reply:
Yea, so, for what it's worth.  On Python 2.7, `bytes` is an alias for the type that calls itself `str` and historically was referred to as `str`.  This is coherent because the Python 2.7 `str` type *is* a byte-oriented type.  It is a sequence of "characters" in the range 0-255.  On Python 3.x, that type has been renamed `bytes` (so `bytes` is no longer just an alias for it, it is actually the name) - also various other changes were made to the type itself that aren't directly relevant here.  As before, it is still a sequence of "characters" in the range 0-255 but it is now preferred to think of it as a sequence of one byte integer values instead of "characters".  Also on Python 3.x, the type that was `unicode` on Python 2.7 has been renamed `str`.

Thus, `str` means one thing on Python 2.7 and something different and very incompatible on Python 3.x.  However, `bytes` means more or less the same thing on each version.  Unfortunately, there is no `unicode` alias on Python 3.x so there is no good unambigious way to refer to that type across both versions.  However, the `future` library offers a `unicode` type so it probably still makes sense to use the name "unicode".

There are cases where you want to document the type of something as `str` but these are the cases where the type itself depends on whether you're on Python 2.7 or Python 3.x.  For example, the type of `""` is `str` on both versions because the meaning of the literal string syntax was changed.  There are also some libraries that will change the type they accept or return based on Python version.  In those cases, `str` is the right thing to use in documentation.  In other cases, `bytes` makes more sense.

Reply:
Thanks for taking the time to explain this, @exarkun!  This is easier for me to understand than what I've seen in the official documentation, but even then I suspect I am going to need some time (and maybe some scars) to actually internalize this. :-)